simulator/resourcewatcher/resourcewatcher.go

```diff
@@ -0,0 +1,323 @@
+		newresourceEventProxy(sw, s.client.StorageV1().RESTClient(), Scs, &storagev1.StorageClass{}, lrVersions.Scs),
+		newresourceEventProxy(sw, s.client.SchedulingV1().RESTClient(), Pcs, &schedulingv1.PriorityClass{}, lrVersions.Pcs),
+	}
+	runctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for _, p := range proxies {
+		go s.Run(p, runctx.Done(), cancel)
+	}
+
+	select {
+	case <-runctx.Done():
+		// runctx monitors s.Run (ListAndWatch) for each resource.
+		// If some error occurs in the process before starting the watch,
+		// this error is returned.
+		return xerrors.Errorf("failed to run ListAndWatch: %w", runctx.Err())
+	case <-ctx.Done():
+		// This method will return an error and finish to event send when the connection from a client is closed.
+		// It includes browser reload, ReadableStream.cancel() calling and so on.
+		// This is to allow the front end to handle stream connection disconnections.
+		return nil
+	}
+}
+
+// Run runs ListAndWatch method.
+// If an error is returned, call cancel to abort ListAndWatch of other resources being processed in parallel.
+func (s *Service) Run(p *resourceEventProxy, stopCh <-chan struct{}, cancel context.CancelFunc) {
+	defer cancel()
+	// ListAndWatch usually continues to wait for WATCH to end and does not return any value.
+	if err := s.ListAndWatch(p, stopCh); err != nil {
+		cancel()
+		klog.Errorf("call ListAndWatch: %w", err)
+	}
+}
+
+// ListAndWatch runs list and watch on the target resource. The list is not always ran
+// This method returns error unless an error occurs in the watch. If an error occurs in the watch,
+// it outputs a log and re-run the watch.
+func (s *Service) ListAndWatch(p ResourceEventProxy, stopCh <-chan struct{}) error {
+	lw := createListWatch(p)
+	// If the lastResourceVersion isn't specified by client, call the list and return the result as ADDED event first.
+	if p.LastResourceVersion() == "" {
+		if err := p.ListAndHandleItems(lw); err != nil {
+			return xerrors.Errorf("call to ListAndHandleItems of %s: %w", p.ResourceKind(), err)
+		}
+	}
+	watcher, err := createWatcher(p, lw)
+	if err != nil {
+		return xerrors.Errorf("call to createWatcher of %s: %w", p.ResourceKind(), err)
+	}
+	p.WatchAndHandleEvent(watcher, stopCh)
+	return nil
+}
+
+// createListWatch creates and returns ListWatch.
+func createListWatch(p ResourceEventProxy) cache.ListerWatcher {
+	return cache.NewListWatchFromClient(p.RestClient(), string(p.ResourceKind()), corev1.NamespaceAll, fields.Everything())
+}
+
+// createWatcher creates and returns RetryWatcher.
+func createWatcher(p ResourceEventProxy, lw cache.ListerWatcher) (watch.Interface, error) {
+	rWatcher, err := watchtools.NewRetryWatcher(p.LastResourceVersion(), lw)
+	if err != nil {
+		return nil, xerrors.Errorf("call NewRetryWatcher: %w", err)
+	}
+	return rWatcher, nil
+}
+
+// ResourceEventProxy is an interface that allows handle events and errors.
```

Comment:
move ResourceEventProxy related definitions(func, struct, and interface) to another file in the same package for visibility.